From 369d9680742be8f51a00f2b58df772df3870d45c Mon Sep 17 00:00:00 2001
From: Mirela Rabulea <mirela.rabulea@nxp.com>
Date: Sun, 15 Sep 2019 23:39:34 +0300
Subject: [PATCH 04/37] MLK-22592: mxc-jpeg: Fix v4l2 compliance, "Format
 ioctls", TRY_FMT, limits

commit 00a3050612f2bba57bb7d1b1c4a97c1d2952e471 from
https://source.codeaurora.org/external/imx/linux-imx.git

Fail: v4l2-test-formats.cpp(476): tot_bytesperline &&
      tot_bytesperline < pix_mp.width
Fix: Do not let mxc_jpeg_try_fmt accept w/h larger than max

Fail (decoder only): v4l2-test-formats.cpp(473): !pfmt.sizeimage
Fix: limit input jpeg stream to the max value, before attempting to align,
     to avoid overflow. Also, remove mxc_jpeg_align, use ALIGN, as it
     does the same thing.
     Input jpeg stream size must be aligned to 1024, not 128

Signed-off-by: Mirela Rabulea <mirela.rabulea@nxp.com>
Reviewed-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Reviewed-by: Robert Chiras <robert.chiras@nxp.com>
(cherry picked from commit 29a870b3b2ffe06759feea9b0aff7dcb4918f07c)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-jpeg.c | 23 +++++++++++------------
 drivers/media/platform/imx8/mxc-jpeg.h |  1 +
 2 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/drivers/media/platform/imx8/mxc-jpeg.c b/drivers/media/platform/imx8/mxc-jpeg.c
index 68efd64e5409..56d98609b522 100644
--- a/drivers/media/platform/imx8/mxc-jpeg.c
+++ b/drivers/media/platform/imx8/mxc-jpeg.c
@@ -288,11 +288,6 @@ static void print_nbuf_to_eoi(struct device *dev, struct vb2_buffer *buf, int n)
 	kfree(bufstr);
 }
 
-static inline u32 mxc_jpeg_align(u32 val, u32 align)
-{
-	return (val + align - 1) & ~(align - 1);
-}
-
 static inline struct mxc_jpeg_ctx *mxc_jpeg_fh_to_ctx(struct v4l2_fh *fh)
 {
 	return container_of(fh, struct mxc_jpeg_ctx, fh);
@@ -780,7 +775,7 @@ static void mxc_jpeg_config_dec_desc(struct vb2_buffer *out_buf,
 		cfg_desc->line_pitch = MXC_JPEG_MIN_WIDTH * 2;
 		cfg_desc->stm_ctrl = STM_CTRL_IMAGE_FORMAT(MXC_JPEG_YUV422);
 		cfg_desc->stm_bufbase = cfg_stream_handle;
-		cfg_desc->stm_bufsize = mxc_jpeg_align(*cfg_size, 1024);
+		cfg_desc->stm_bufsize = ALIGN(*cfg_size, 1024);
 		print_descriptor_info(jpeg->dev, cfg_desc);
 	}
 
@@ -794,8 +789,7 @@ static void mxc_jpeg_config_dec_desc(struct vb2_buffer *out_buf,
 	desc->line_pitch = q_data_cap->bytesperline[0];
 
 	mxc_jpeg_addrs(desc, dst_buf, src_buf, 0);
-	mxc_jpeg_set_bufsize(desc,
-			mxc_jpeg_align(vb2_plane_size(src_buf, 0), 1024));
+	mxc_jpeg_set_bufsize(desc, ALIGN(vb2_plane_size(src_buf, 0), 1024));
 	print_descriptor_info(jpeg->dev, desc);
 	if (ctx->dht_needed) {
 		/* validate the configuration descriptor */
@@ -1616,8 +1610,11 @@ static int mxc_jpeg_try_fmt(struct v4l2_format *f, struct mxc_jpeg_fmt *fmt,
 {
 	struct v4l2_pix_format_mplane *pix_mp = &f->fmt.pix_mp;
 	struct v4l2_plane_pix_format *pfmt;
-	u32 w = pix_mp->width;
-	u32 h = pix_mp->height;
+	u32 w = (pix_mp->width < MXC_JPEG_MAX_WIDTH) ?
+		 pix_mp->width : MXC_JPEG_MAX_WIDTH;
+	u32 h = (pix_mp->height < MXC_JPEG_MAX_HEIGHT) ?
+		 pix_mp->height : MXC_JPEG_MAX_HEIGHT;
+
 	unsigned int mode = ctx->mode;
 	int i;
 	bool is_nv12 = (fmt->fourcc == V4L2_PIX_FMT_NV12);
@@ -1653,10 +1650,12 @@ static int mxc_jpeg_try_fmt(struct v4l2_format *f, struct mxc_jpeg_fmt *fmt,
 		if (q_type == MXC_JPEG_FMT_TYPE_ENC &&
 		    mode == MXC_JPEG_DECODE) {
 			pfmt->bytesperline = 0;
-			/* Source size must be aligned to 128 */
-			pfmt->sizeimage = mxc_jpeg_align(pfmt->sizeimage, 128);
 			if (pfmt->sizeimage == 0)
 				pfmt->sizeimage = MXC_JPEG_DEFAULT_SIZEIMAGE;
+			if (pfmt->sizeimage > MXC_JPEG_MAX_SIZEIMAGE)
+				pfmt->sizeimage = MXC_JPEG_MAX_SIZEIMAGE;
+			/* input jpeg stream must be aligned to 1024 */
+			pfmt->sizeimage = ALIGN(pfmt->sizeimage, 1024);
 		} else if (q_type == MXC_JPEG_FMT_TYPE_RAW &&
 		    mode == MXC_JPEG_DECODE) {
 			pfmt->bytesperline = w * (fmt->depth / 8);
diff --git a/drivers/media/platform/imx8/mxc-jpeg.h b/drivers/media/platform/imx8/mxc-jpeg.h
index d7bba8ce08c4..3be9b5753d3b 100644
--- a/drivers/media/platform/imx8/mxc-jpeg.h
+++ b/drivers/media/platform/imx8/mxc-jpeg.h
@@ -35,6 +35,7 @@
 #define MXC_JPEG_W_ALIGN		3
 #define MXC_JPEG_DEFAULT_SIZEIMAGE	(6 * MXC_JPEG_DEFAULT_WIDTH * \
 					 MXC_JPEG_DEFAULT_HEIGHT)
+#define MXC_JPEG_MAX_SIZEIMAGE		0xFFFFFC00
 #define MXC_JPEG_ENC_CONF		1
 #define MXC_JPEG_ENC_DONE		0
 #define SOF0				0xC0
-- 
2.17.1

